proxy_ring_buffer.go

```diff
@@ -0,0 +1,278 @@
+	// going to overwrite that item for the next 29 seconds.
+	OverflowPolicyFail OverflowPolicy = 1
+
+	// ReadResultSetSequenceUnavailable is used when error happened
+	ReadResultSetSequenceUnavailable int64 = -1
+)
+
+func newRingbuffer(p *proxy) (*Ringbuffer, error) {
+	partitionID, err := p.stringToPartitionID(p.name)
+	if err != nil {
+		return nil, err
+	}
+	return &Ringbuffer{proxy: p, partitionID: partitionID, capacity: ReadResultSetSequenceUnavailable}, nil
+}
+
+// Add an item to the tail of the Ringbuffer. If there is space in the Ringbuffer, the call
+// will return the sequence of the written item. If there is no space, it depends on the overflow policy what happens:
+// OverflowPolicy OverflowPolicyOverwrite  we just overwrite the oldest item in the Ringbuffer, and we violate the ttl
+// OverflowPolicy FAIL we return -1. The reason that FAIL exist is to give the opportunity to obey the ttl.
+//
+// This sequence will always be unique for this Ringbuffer instance, so it can be used as a unique id generator if you are
+// publishing items on this Ringbuffer. However, you need to take care of correctly determining an initial id when any node
+// uses the Ringbuffer for the first time. The most reliable way to do that is to write a dummy item into the Ringbuffer and
+// use the returned sequence as initial  id. On the reading side, this dummy item should be discarded. Please keep in mind that
+// this id is not the sequence of the item you are about to publish but from a previously published item. So it can't be used
+// to find that item.
+//
+// Add returns the sequence number of the added item. You can read the added item using this number.
+func (rb *Ringbuffer) Add(ctx context.Context, item interface{}, overflowPolicy OverflowPolicy) (sequence int64, err error) {
+	elementData, err := rb.validateAndSerialize(item)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	request := codec.EncodeRingbufferAddRequest(rb.name, int32(overflowPolicy), elementData)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferAddResponse(response), nil
+}
+
+// AddAll
+// Adds all the items of a collection to the tail of the Ringbuffer. A addAll is likely to outperform multiple calls
+// to add(Object) due to better io utilization and a reduced number of executed operations. If the batch is empty,
+// the call is ignored. When the collection is not empty, the content is copied into a different data-structure.
+// This means that: after this call completes, the collection can be re-used. the collection doesn't need to be serializable.
+// If the collection is larger than the capacity of the Ringbuffer, then the items that were written first will be
+// overwritten. Therefor this call will not block. The items are inserted in the order of the Iterator of the collection.
```

Comment:
> The items are inserted in the order of the Iterator of the collection

This looks like a Java specific sentence. In general can you refactor whole comment so that it does not mention a collection but a slice maybe?

Reply:
Done.
ff4bf6521c968ebf1ce850a4ef29a818c7c156e7